aws/transport/http/request_id.go

```diff
@@ -0,0 +1,40 @@
+package http
+
+import (
+	"fmt"
+
+	"github.com/awslabs/smithy-go/middleware"
+)
+
+// requestID is used to retrieve request id from response metadata
+type requestID struct {
+}
+
+// SetRequestIDMetadata sets the provided request id over middleware metadata
+func SetRequestIDMetadata(metadata *middleware.Metadata, id string) {
+	metadata.Set(requestID{}, id)
+}
+
+// GetRequestIDMetadata retrieves the request id from middleware metadata
+func GetRequestIDMetadata(metadata middleware.Metadata) string {
+	if !metadata.Has(requestID{}) {
+		return ""
+	}
+
+	v, ok := metadata.Get(requestID{}).(string)
+	if !ok {
+		return ""
+	}
+	return v
+}
+
+// SetRequestIDMetadataFromReader takes in dst metadata on which metadata needs to be set,
+// and a src metadata reader from which the metadata is read.
+func SetRequestIDMetadataFromReader(dst *middleware.Metadata, src middleware.MetadataReader) error {
+	id, ok := src.Get(requestID{}).(string)
+	if !ok {
+		return fmt.Errorf("expected requestID to present as string in middleware metadata")
+	}
+	SetRequestIDMetadata(dst, id)
+	return nil
```

Comment:
A reader wouldn't be needed if the deserialization function took `*middleware.Metadata` as input. This would also remove the need for the additional `middleware.Metadata` `deserMetadata` return value.